ResearchKit/Common/ORKStepNavigationRule.h

```diff
@@ -0,0 +1,371 @@
+ 
+ It can be used to match any answer combination in the results of the ongoing task and jump accordingly. You must provide one or more result predicates (each predicate can match one or more step result within the task).
+ 
+ The `ORKResultPredicate` class provides convenience class methods to build predicates for all the `ORKQuestionResult` subtypes.
+ */
+ORK_CLASS_AVAILABLE
+@interface ORKPredicateStepNavigationRule : ORKStepNavigationRule
+
+/**
+ Returns an initialized predicate step navigation rule using the specified result predicates, matching step identifiers, and an optional default step identifier.
+ 
+ @param resultPredicates            An array of result predicates. Each result predicate can match one or more step results in the ongoing task.
+ @param matchingStepIdentifiers     An array of possible destination step identifiers. This array must contain one step identifier for each of the predicates in `resultPredicates`.
+ @param defaultStepIdentifier       The identifier of the step which will be used if none of the result predicates match. If this argument is `nil` and none of the predicates match, the default ordered task navigation behavior takes place (i.e, the task goes to the next step in order).
+ 
+ @return An initialized predicate step navigation rule.
+ */
+- (instancetype)initWithResultPredicates:(NSArray *)resultPredicates
+                 matchingStepIdentifiers:(NSArray *)matchingStepIdentifiers
+                   defaultStepIdentifier:(nullable NSString *)defaultStepIdentifier NS_DESIGNATED_INITIALIZER;
+
+/**
+ Returns an initialized predicate step navigation rule using the specified result predicates and matching step identifiers.
+ 
+ @param resultPredicates            An array of result predicates. Each result predicate can match one or more step results in the ongoing task.
+ @param matchingStepIdentifiers     An array of possible destination step identifiers. This array must contain one step identifier for each of the predicates in resultPredicates.
+ 
+ @return An initialized predicate step navigation rule.
+ */
+- (instancetype)initWithResultPredicates:(NSArray *)resultPredicates
+                 matchingStepIdentifiers:(NSArray *)matchingStepIdentifiers;
+
+/**
+ Returns a new predicate step navigation rule initialized from data in the given unarchiver.
+ 
+ @param aDecoder    Coder from which to initialize the step navigation rule.
+ @return A new predicate step navigation rule.
+ */
+- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_DESIGNATED_INITIALIZER;
+
+/**
+ An optional array of additional task results.
+ 
+ The predicate step navigation rule will use the child step results within these tasks, in addition to the current task results, to math the result predicates. You must ensure that all the step identifiers within the additional task results are unique, and that they are different from the step identifiers within the current task result.
```

Comment:
It's useful if you want to add conditional navigation logic to your task, based on the results on one or more previously completed task. E.g., if the participant answers an "About you" task the first day of study, you could then have other tasks with common parts and specific questions for different participant parameters.

This is not currently used in the `ORKTest`'s interactive example, but you can see a small non-interactive example in the `testPredicateStepNavigationRule` method of `ORKTaskTest`.